### PR TITLE
[Core] BDF-schemes - using time discretization utility

### PR DESCRIPTION
**Pull request by ddiezrod**
BDF schemes use now the time discretization utility to compute the bdf coefficients.

### CODE REVIEW

File: kratos/solving_strategies/schemes/residual_based_bdf_scheme.h [hunk 1]
```diff
@@ -383,9 +327,11 @@ class ResidualBasedBDFScheme
         std::vector< Vector > dot2un0; /// Second derivative
     };
 
-    const std::size_t mOrder; /// The integration order
-    Vector mBDF;              /// The BDF coefficients
-    GeneralVectors mVector;   /// The structure containing the  derivatives
+    const std::size_t mOrder;                      /// The integration order
+    Vector mBDF;                                   /// The BDF coefficients
+    GeneralVectors mVector;                        /// The structure containing the  derivatives
+    TimeDiscretization::BDF::Pointer mpBDFUtility; /// Utility to compute BDF coefficients
```
Comment (philbucher) [inline kratos/solving_strategies/schemes/residual_based_bdf_scheme.h]: I don't see the need for this to be a pointer

Comment (ddiezrod) [reply kratos/solving_strategies/schemes/residual_based_bdf_scheme.h]: you are right, the habit... i will change it

File: kratos/solving_strategies/schemes/residual_based_bdf_scheme.h [hunk 2]
```diff
@@ -242,66 +244,8 @@ class ResidualBasedBDFScheme
...
-
-        const double tolerance = 1.0e-24;
-        KRATOS_WARNING_IF("ResidualBasedBDFScheme", mOrder > 2 && std::abs(delta_time - previous_delta_time) > tolerance) << "For higher orders than 2 the time step is assumed to be constant. Sorry for the inconveniences" << std::endl;
-
-        // Adding to the process info
-        Vector bdf_vector(mOrder + 1);
-        for (std::size_t i_order = 0; i_order < mOrder + 1; ++i_order)
-            bdf_vector[i_order] = mBDF[i_order];
-        r_current_process_info.SetValue(BDF_COEFFICIENTS, bdf_vector);
+        mpBDFUtility->ComputeAndSaveBDFCoefficients(r_current_process_info);
```
Comment (loumalouomega) [inline kratos/solving_strategies/schemes/residual_based_bdf_scheme.h:249]: Does the utilities compute the coefficients to order greater than 2?

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Please, if order greater than 2 is not computed leave that part of the code

File: kratos/solving_strategies/schemes/residual_based_bdf_scheme.h (hunk 2 above)
Comment (philbucher) [reply kratos/solving_strategies/schemes/residual_based_bdf_scheme.h:249]: [It does](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/utilities/time_discretization.cpp#L33-L50)

**Comment by philbucher** [comment]
@ddiezrod please wait with merging for the other PR, as this might also reveal some problems, then it will take longer

File: kratos/solving_strategies/schemes/residual_based_bdf_scheme.h (hunk 2 above)
Comment (ddiezrod) [reply kratos/solving_strategies/schemes/residual_based_bdf_scheme.h:249]: @loumalouomega you mean you want to maintain the warning if time order > 2? Maybe we can move it to the time discretization utility...

File: kratos/solving_strategies/schemes/residual_based_bdf_scheme.h (hunk 1 above)
Comment (ddiezrod) [reply kratos/solving_strategies/schemes/residual_based_bdf_scheme.h]: I think it needs to be a pointer. Time discretization does not have copy constructor (it has a unique pointer inside) and time scheme class needs it.

File: kratos/solving_strategies/schemes/residual_based_bdf_scheme.h [hunk 3]
```diff
@@ -121,7 +122,8 @@ class ResidualBasedBDFScheme
      */
     explicit ResidualBasedBDFScheme(const std::size_t Order = 2)
         :ImplicitBaseType(),
-         mOrder(Order)
+         mOrder(Order),
+         mpBDFUtility(Kratos::make_shared<TimeDiscretization::BDF>(Order))
```
Comment (philbucher) [inline kratos/solving_strategies/schemes/residual_based_bdf_scheme.h]: Ok if it has to be a pointer then pls use unique

```suggestion
         mpBDFUtility(Kratos::make_unique<TimeDiscretization::BDF>(Order))
```

File: kratos/solving_strategies/schemes/residual_based_bdf_scheme.h (hunk 2 above)
Comment (loumalouomega) [reply kratos/solving_strategies/schemes/residual_based_bdf_scheme.h:249]: Yep, that will be fine

File: kratos/solving_strategies/schemes/residual_based_bdf_scheme.h (hunk 1 above)
Comment (rubenzorrilla) [reply kratos/solving_strategies/schemes/residual_based_bdf_scheme.h]: And if you make it a pointer we can use any BDF order (1,2,3..)

File: kratos/solving_strategies/schemes/residual_based_bdf_scheme.h (hunk 2 above)
Comment (ddiezrod) [reply kratos/solving_strategies/schemes/residual_based_bdf_scheme.h:249]: warning recovered in scheme

File: kratos/solving_strategies/schemes/residual_based_bdf_scheme.h [hunk 4]
```diff
@@ -238,70 +241,15 @@ class ResidualBasedBDFScheme
...
+        mBDF = r_current_process_info[BDF_COEFFICIENTS];
 
-        // Adding to the process info
-        Vector bdf_vector(mOrder + 1);
-        for (std::size_t i_order = 0; i_order < mOrder + 1; ++i_order)
-            bdf_vector[i_order] = mBDF[i_order];
-        r_current_process_info.SetValue(BDF_COEFFICIENTS, bdf_vector);
+        KRATOS_WARNING_IF("ResidualBasedBDFScheme", mOrder > 2)
+        << "For higher orders than 2 the time step is assumed to be constant."
+        << "Sorry for the inconveniences \n";
```
Comment (rubenzorrilla) [inline kratos/solving_strategies/schemes/residual_based_bdf_scheme.h]:
```suggestion
```

File: kratos/solving_strategies/schemes/residual_based_bdf_scheme.h [hunk 5]
```diff
@@ -238,70 +241,15 @@ class ResidualBasedBDFScheme
...
+        mpBDFUtility->ComputeAndSaveBDFCoefficients(r_current_process_info);
+        mBDF = r_current_process_info[BDF_COEFFICIENTS];
 
-        // Adding to the process info
-        Vector bdf_vector(mOrder + 1);
-        for (std::size_t i_order = 0; i_order < mOrder + 1; ++i_order)
-            bdf_vector[i_order] = mBDF[i_order];
-        r_current_process_info.SetValue(BDF_COEFFICIENTS, bdf_vector);
+        KRATOS_WARNING_IF("ResidualBasedBDFScheme", mOrder > 2)
+        << "For higher orders than 2 the time step is assumed to be constant."
```
Comment (rubenzorrilla) [inline kratos/solving_strategies/schemes/residual_based_bdf_scheme.h]:
```suggestion
        << "For higher orders than 2 the time step is assumed to be constant.";
```

Comment (ddiezrod) [reply kratos/solving_strategies/schemes/residual_based_bdf_scheme.h]: That would cause an error - notice that message continues in following line.

Comment (ddiezrod) [reply kratos/solving_strategies/schemes/residual_based_bdf_scheme.h]: Ok, I undestand your point now, you want to remove the "sorry, not really sorry" part. Fine for me.

File: kratos/solving_strategies/schemes/residual_based_bdf_scheme.h (hunk 1 above)
Comment (philbucher) [reply kratos/solving_strategies/schemes/residual_based_bdf_scheme.h]: ok I get it now :)

Comment (philbucher) [reply kratos/solving_strategies/schemes/residual_based_bdf_scheme.h]: btw does this even compile?
I thought you changed to unique ptr?

Comment (ddiezrod) [reply kratos/solving_strategies/schemes/residual_based_bdf_scheme.h]: My bad!, I think it is actually still a shared pointer. If I change this to:

`Kratos::unique_ptr<TimeDiscretization::BDF> mpBDFUtility;`

I get an error in the copy constructor, so I will leave it as a shared pointer (using make_shared)

Comment (philbucher) [reply kratos/solving_strategies/schemes/residual_based_bdf_scheme.h]: ok

File: kratos/solving_strategies/schemes/residual_based_bdf_scheme.h [hunk 6]
```diff
@@ -123,7 +123,7 @@ class ResidualBasedBDFScheme
     explicit ResidualBasedBDFScheme(const std::size_t Order = 2)
         :ImplicitBaseType(),
          mOrder(Order),
-         mpBDFUtility(Kratos::make_unique<TimeDiscretization::BDF>(Order))
```
Comment (rubenzorrilla) [inline kratos/solving_strategies/schemes/residual_based_bdf_scheme.h]: I think that it is possible to use a unique pointer if you create an auxiliary unique pointer and then set the member one value by doing a `swap` operation.

Comment (ddiezrod) [reply kratos/solving_strategies/schemes/residual_based_bdf_scheme.h]: I can do this guys if you think its better than having a shared pointer. I really do not mind. @philbucher

Comment (rubenzorrilla) [reply kratos/solving_strategies/schemes/residual_based_bdf_scheme.h]: Well, it's just a pointer. But in this case makes sense that it is unique (there will be only one BDF utils). Moreover a unique pointer is much cheaper.

Comment (ddiezrod) [reply kratos/solving_strategies/schemes/residual_based_bdf_scheme.h]: Done, please check it is correct @rubenzorrilla

**Comment by ddiezrod** [comment]
Is this ok for merging? @loumalouomega @philbucher @rubenzorrilla

**Comment by loumalouomega** [comment]
Can you add my name to the authorship of TimeDiscrtization utility?, some of my code was refactored and I think it is fare

**Comment by philbucher** [comment]
Please merge master, then I will run some of my tests to check that nothing breaks

**Comment by philbucher** [comment]
> Can you add my name to the authorship of TimeDiscrtization utility?, some of my code was refactored and I think it is fare

I will not block bcs of this but I don’t think there is anything refactored, I rewrote it from scratch 
Which part of the code are you referring to?

**Comment by rubenzorrilla** [comment]
@ddiezrod the changes make sense for me. I let @philbucher and @loumalouomega to give you the final approve.

**Review by loumalouomega — APPROVED** [review]
OK

**Review by philbucher — APPROVED** [review]
thanks @ddiezrod

**Comment by ddiezrod** [comment]
Thanks all!

**Pull request merged**